### PR TITLE
jenkins-plugin: properly handle cancelled jobs

### PR DESCRIPTION
**Pull request by gabemontero**
Bug 1357752
Bugzilla link https://bugzilla.redhat.com/show_bug.cgi?id=1357752
Detailed explanation
The pipeline plugin needed to throw InterruptedExceptions instead of
swallowing them in order to properly handle cancelled jobs

### CODE REVIEW

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftBuilder.java [hunk 1]
```diff
@@ -163,14 +164,16 @@ public void onFailure(IOException e) {
             try {
                 Thread.sleep(1000);
             } catch (InterruptedException e) {
+                // need to throw as this indicates the step as been cancelled
+                throw e;
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftBuilder.java:173]: we should also invoke a cancel build action to cancel the build running on openshift.

ditto for the deployment step.

**Review by bparees — CHANGES_REQUESTED** [review]

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftBuilder.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftBuilder.java:173]: I'll investigate

**Comment by gabemontero** [comment]
Also note - I've updated the PR tester job with @jupierce 's new PR testing changes ... so vetting that piece with this PR as well.

**Comment by gabemontero** [comment]
Note - manual testing based on bugzilla scenario looks good, including the cancel of the app build per suggestion from @bparees 

Jenkins job output:

```
Starting the "Trigger OpenShift Build" step with build config "nodejs-mongodb-example" from the project "myproject".
  Started build "nodejs-mongodb-example-1" and waiting for build completion ...
Operation will timeout after 900000 milliseconds
Aborted by anonymous


Starting the "Cancel OpenShift Builds" step with build config "nodejs-mongodb-example" from the project "myproject".
Aborted by anonymous
  Cancelled build "nodejs-mongodb-example-1".


Exiting "Cancel OpenShift Builds" successfully with 1 builds cancelled.
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Aborted by anonymous
Aborted by anonymous
```

Output from `oc get builds`:

```
gmontero ~/go/src/github.com/openshift/origin  (master)$ oc get builds
NAME                       TYPE              FROM          STATUS      STARTED              DURATION
nodejs-mongodb-example-1   Source            Git           Cancelled   20 seconds ago       6s
sample-pipeline-1          JenkinsPipeline                 Cancelled   About a minute ago   51s
```

Regression testing of the freestyle sample job also passed.

Lastly, the update to the PR test job looks good.  We are calling @jupierce 's new script and passing the new env var to the extended test invocation and things got invoked correctly.  Some of the known issues with the extended test are still out there waiting to get their fixes merged (https://github.com/openshift/origin/pull/11711)

Unless we have more real fixes required, I'll do a test PR to verify everything is in order once all the origin side dependencies have merged.

Will push updates momentarily, allow some time for review before merging and cutting next level for QE to verify.

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftBuilder.java [hunk 2]
```diff
@@ -250,6 +258,8 @@ default boolean coreLogic(Launcher launcher, TaskListener listener, Map<String,
                         try {
                             Thread.sleep(1000);
                         } catch (InterruptedException e) {
+                            // need to throw as this indicates the step as been cancelled
+                            throw e;
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftBuilder.java:266]: wouldn't we cancel the build here too?

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftDeployer.java [hunk 3]
```diff
@@ -112,6 +112,8 @@ public IDeploymentConfig visit(IDeploymentTriggerable triggerable) {
                         try {
                             Thread.sleep(10000);
                         } catch (InterruptedException e) {
+                            // need to throw as this indicates the step as been cancelled
+                            throw e;
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftDeployer.java:121]: and cancel the deployment here also?

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftPlugin.java [hunk 4]
```diff
@@ -338,6 +341,8 @@ default boolean verifyBuild(long startTime, long wait, IClient client, String bl
                 try {
                     Thread.sleep(1000);
                 } catch (InterruptedException e) {
+                    // need to throw as this indicates the step as been cancelled
+                    throw e;
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftPlugin.java:350]: cancel build here?

**Comment by gabemontero** [comment]
additional cancels pushed

**Comment by bparees** [comment]
lgtm

**Pull request merged**